src/System.ServiceProcess.ServiceController/tests/ServiceBaseTests.cs

```diff
@@ -73,106 +76,140 @@ public void Cleanup()
         }
 
         [ConditionalFact(nameof(IsProcessElevated))]
-        public void TestOnStartThenStop()
+        public async Task TestOnStartThenStopAsync()
         {
+            Task readTask;
+            byte[] received = new byte[] { 0 };
+            _testService.Client.Connect(connectionTimeout);
             var controller = new ServiceController(_testService.TestServiceName);
             AssertExpectedProperties(controller);
-            string expected =
-@"OnStart args=
-OnStop
-";
+
             controller.Stop();
+            readTask = _testService.Client.ReadAsync(received, 0, 1);
+            await readTask.TimeoutAfter(readTimeout);
+            Assert.Equal((int)PipeMessageByteCode.Stop, received[0]);
             controller.WaitForStatus(ServiceControllerStatus.Stopped);
-            Assert.Equal(expected, _testService.GetServiceOutput());
         }
 
         [ConditionalFact(nameof(IsProcessElevated))]
-        public void TestOnStartWithArgsThenStop()
+        public async Task TestOnStartWithArgsThenStopAsync()
         {
+            Task readTask;
+            byte[] received = new byte[] { 0 };
             var controller = new ServiceController(_testService.TestServiceName);
+            _testService.Client.Connect(connectionTimeout);
             AssertExpectedProperties(controller);
+
             controller.Stop();
+            readTask = _testService.Client.ReadAsync(received, 0, 1);
```

Comment:
A lot of this code is still duplicated and can be encapsulated more in the _testService. Ideally I think the test code should looks something like:

```
using (_testService.ConnectToClient())
{
... do work
byte[] received = _testService.ReadBytes(). 
}
```

Reply:
That is of course if you can share the same client. If you cannot share the same client then you should return a unique client that can get you the information.

Reply:
```
        public void TestOnExecuteCustomCommandAsync()
        {
            _testService.Client.Connect(connectionTimeout);
            var controller = new ServiceController(_testService.TestServiceName);
            AssertExpectedProperties(controller);

            controller.ExecuteCommand(128);
            var task = Task.Run(() => _testService.ReadPipeAsync());
            task.Wait();
            Assert.Equal(128, task.Result);

            controller.Stop();
            task = Task.Run(() => _testService.ReadPipeAsync());
            task.Wait();
            Assert.Equal((int)PipeMessageByteCode.Stop, task.Result);
            controller.WaitForStatus(ServiceControllerStatus.Stopped);
        }
```

Reply:
That is better but I still think we can do better. We also need to be sure that we correctly clean-up the pipe client. 

Why do you need to to the extra task work in every test method? That should be encapsulated.

Reply:
```
            _testService.Client.Connect(connectionTimeout);
            var controller = new ServiceController(_testService.TestServiceName);
            AssertExpectedProperties(controller);

            controller.ExecuteCommand(128);
            Assert.Equal(128, GetBytes());

            controller.Stop();
            Assert.Equal((int)PipeMessageByteCode.Stop, GetBytes());
            controller.WaitForStatus(ServiceControllerStatus.Stopped);
``

Reply:
That is better. I think I would still put the GetBytes method on _testService though.

Reply:
Much better. Now the only question for me is whether or not you need to clean up the client during the test execution at all. Should it happen as part of your GetBytes call, or can you call GetBytes multiple times with the same open client?

Reply:
yeah I can call getbytes multiple times with the same open client. infact this is the case with the most tests.